x-pack/metricbeat/module/oracle/tablespace/data.go

```diff
@@ -0,0 +1,156 @@
+		return nil, errors.Wrap(err, "error getting temp_free_space")
+	}
+
+	if out.freeSpace, err = extractor.usedAndFreeSpaceData(); err != nil {
+		return nil, errors.Wrap(err, "error getting free space data")
+	}
+
+	return
+}
+
+// transform is the T of an ETL (refer to the 'extract' method above if you need to see the origin). Transforms the data
+// to create a Kibana/Elasticsearch friendly JSON. Data from Oracle is pretty fragmented by design so a lot of data
+// was necessary. Data is organized by Tablespace entity (Tablespaces might contain one or more data files)
+func (m *MetricSet) transform(in *extractedData) (out map[string]common.MapStr) {
+	out = make(map[string]common.MapStr, 0)
+
+	for _, dataFile := range in.dataFiles {
+		m.addDataFileData(&dataFile, out)
+	}
+
+	m.addUsedAndFreeSpaceData(in.freeSpace, out)
+	m.addTempFreeSpaceData(in.tempFreeSpace, out)
+
+	return
+}
+
+func (m *MetricSet) eventMapping() ([]mb.Event, error) {
+	extractedMetricsData, err := m.extract(m.extractor)
+	if err != nil {
+		return nil, errors.Wrap(err, "error extracting data")
+	}
+
+	out := m.transform(extractedMetricsData)
+
+	events := make([]mb.Event, 0)
+	for _, v := range out {
+		events = append(events, mb.Event{MetricSetFields: v})
+	}
+
+	return events, nil
+}
+
+// addTempFreeSpaceData is specific to the TEMP Tablespace.
+func (m *MetricSet) addTempFreeSpaceData(tempFreeSpaces []tempFreeSpace, out map[string]common.MapStr) {
+	for key, cm := range out {
+		val, err := cm.GetValue("name")
+		if err != nil {
+			m.Logger().Debug("error getting tablespace name")
+			continue
+		}
+		name := val.(string)
+		if name == "TEMP" {
+			for _, tempFreeSpaceTable := range tempFreeSpaces {
+				m.checkNullInt64(out, key, "space.total.bytes", tempFreeSpaceTable.TablespaceSize)
+				m.checkNullInt64(out, key, "space.used.bytes", tempFreeSpaceTable.UsedSpaceBytes)
+				m.checkNullInt64(out, key, "space.free.bytes", tempFreeSpaceTable.FreeSpace)
+			}
+		}
+	}
+}
+
+// addUsedAndFreeSpaceData is specific to all Tablespaces but TEMP
+func (m *MetricSet) addUsedAndFreeSpaceData(freeSpaces []usedAndFreeSpace, out map[string]common.MapStr) {
+	for key, cm := range out {
+		val, err := cm.GetValue("name")
+		if err != nil {
+			m.Logger().Debug("error getting tablespace name")
+			continue
+		}
+		name := val.(string)
+		if name != "" {
+			for _, freeSpaceTable := range freeSpaces {
+				if name == freeSpaceTable.TablespaceName {
+					m.checkNullInt64(out, key, "space.free.bytes", freeSpaceTable.TotalFreeBytes)
+					m.checkNullInt64(out, key, "space.used.bytes", freeSpaceTable.TotalUsedBytes)
+				}
+			}
+		}
+	}
+}
+
+// addDataFileData is a specific data file which generates a JSON output.
+func (m *MetricSet) addDataFileData(res tablespaceNameGetter, output map[string]common.MapStr) {
+	if _, found := output[res.hash()]; !found {
+		output[res.hash()] = common.MapStr{}
+	}
+
+	val, ok := res.(*dataFile)
```

Comment:
this feels wrong, why not passing a `dataFile` directly?

Reply:
True, this method was more complex before and I forgot to rewrite signature.